### PR TITLE
Azure Service Principal with Secret authentication workflow.

### PR DESCRIPTION
**Pull request by kzych-inpost**
## Context

To my best knowledge (version `v1.42.12`), the only way to authentiacate using Service Principal workflow is to provide `azure_ad_token` - either hardcoded (token expires though) or providing a path that starts with `oidc/`: 

```
model_name: "azure-gpt-35"
    litellm_params:
      model: "azure/gpt-35-turbo"
      api_base: "some-url"
      azure_ad_token: "oidc/azure/api://AzureADTokenExchange"
``` 

## Why?

However, this workflow does not suit my needs, as it requires `AZURE_FEDERATED_TOKEN_FILE`. 
My goal is to use [Service Principal with Secret workflow](https://learn.microsoft.com/en-us/python/api/overview/azure/identity-readme?view=azure-python#service-principal-with-secret).

## How? 

There is a possibility in `openai` SDK to use `azure_ad_token_provider` [parameter](https://github.com/openai/openai-python/blob/646fff04dbe9458f1c46753031546ccf6415eeef/src/openai/lib/azure.py#L75), which is not used yet in the `LiteLLM Proxy`.

## Relevant issues

Related issue: https://github.com/BerriAI/litellm/issues/4417.

## Type

🆕 New Feature

## Changes

- introduced `get_azure_ad_token_provider()` function;
- modified `set_client()` so it uses `get_azure_ad_token_provider()` when necessary.

## [REQUIRED] Testing - Attach a screenshot of any new tests passing locall
If UI changes, send a screenshot/GIF of working UI fixes

```
❯ poetry run pytest -q --disable-warnings .
sssssssssssssssssssssssssssssssssssssssssssssssssssssssssss....ssssssssssssssssssss                                           [100%]
4 passed, 79 skipped, 143 warnings in 4.61s
```

### CODE REVIEW

**Comment by kzych-inpost** [comment]
New usage would be simple, without any `azure_ad_token` key:

```
model_name: "azure-gpt-35"
    litellm_params:
      model: "azure/gpt-35-turbo"
      api_base: "some-url"
```

but these environment variables need to be passed: `AZURE_CLIENT_ID`, `AZURE_TENANT_ID`, `AZURE_CLIENT_SECRET`.

**Review by ishaan-jaff — CHANGES_REQUESTED** [review]
hi @kzych-inpost can you send a screenshot of using azure_ad_token_provider working for your on your proxy server ?

**Comment by kzych-inpost** [comment]
Oh right, hope these screenshots would suffice:
 
`docker-compose.yml` - two important lines:
1. line 81 - using image built locally with my change; 
2. lines 94-96 - environment variables: `AZURE_TENANT_ID`, `AZURE_CLIENT_ID` and `AZURE_CLIENT_SECRET` are used by LiteLLM Proxy
Screenshot 2024-08-12 at 10 52 07

`litellm-config.yaml` - defined models from my Azure deployments:
Screenshot 2024-08-12 at 11 03 03

and here is the docker output with successful requests:
Screenshot 2024-08-12 at 10 52 47

**Comment by ghost** [comment]
@ishaan-jaff is this good for merge?

**Comment by shulkx** [comment]
Hopefully, it will be merged soon and then there will no need to refresh the token schedulely anymore when using the model from azure.

**Comment by Manouchehri** [comment]
> However, this workflow does not suit my needs, as it requires `AZURE_FEDERATED_TOKEN_FILE`.

I do not understand this, as `AZURE_FEDERATED_TOKEN_FILE` is _definitely_ already supported, see below. I added this for everyone back in #3861.

https://github.com/BerriAI/litellm/blob/7f39f9f97d2243290bead3c6560c9a9d081119a1/litellm/utils.py#L8428-L8430

What about this doesn't work for you? The example you gave, should functional exactly the same as what this PR.

I'm aware of #4417, which would be nice to have help on. :) 

Could we stick to using the existing `azure_ad_token: "oidc/azure/api://AzureADTokenExchange"` syntax, and instead refactor this to fix #4417?

OIDC is complicated enough, I _really_ don't want to see fragmentation in LiteLLM itself on this.

**Comment by Manouchehri** [comment]
> My goal is to use [Service Principal with Secret workflow](https://learn.microsoft.com/en-us/python/api/overview/azure/identity-readme?view=azure-python#service-principal-with-secret).

This also doesn't make sense to me. Are you trying to use token based Azure AD auth, or client secret auth?

@kzych-inpost: Could you clarify where/what you are running LiteLLM on?

**Comment by Manouchehri** [comment]
This has a _really_ high risk of moving people away from the more secure OIDC flow, and encouraging them to use static secrets instead of more-secure temporary credentials. I would really want to see #4417 fixed before this is merged in.

Also, I really don't like that this is loading way more than the title implies. See this huge list for all the things it's doing:

https://learn.microsoft.com/en-us/dotnet/api/azure.identity.defaultazurecredential?view=azure-dotnet

I would really want to see this narrowed down to just ClientSecretCredential.

https://learn.microsoft.com/en-us/python/api/azure-identity/azure.identity.clientsecretcredential?view=azure-python

And finally, `get_azure_ad_token_provider` is an _incredibly_ misleading name, and will cause people to think that this is the only recommended way of doing auth to Azure, instead of using the more secure (and tested) OIDC flow. At the least, please do not merge this until `get_azure_ad_token_provider` is removed or renamed to something less misleading.

**Review by Manouchehri — CHANGES_REQUESTED** [review]
See comments in https://github.com/BerriAI/litellm/pull/5131#issuecomment-2294215768.

**Comment by kzych-inpost** [comment]
@Manouchehri:

> This has a really high risk of moving people away from the more secure OIDC flow, and encouraging them to use static secrets instead of more-secure temporary credentials.

I see your point, but there are use-cases (like mine - Service Principal authentication), where classic OIDC flow for **user** authentication is not what suits one's need. What is more, the authentication method I've proposed also incorporates fetching **temporary token** (_temporary credential_ mentioned by you), so I believe this method is also quite secure.  


> I would really want to see this narrowed down to just ClientSecretCredential.

Sure, I can do that. 
On the other hand, authenticating using [EnvironmentCredential](https://learn.microsoft.com/en-us/python/api/azure-identity/azure.identity.environmentcredential?view=azure-python) would make Service Principal authentication more flexible - it would be possible not only for `Service principal with secret` flow, but also `Service principal with certificate` and `User with username and password`. What do you think?


> `get_azure_ad_token_provider` is an incredibly misleading name

Well, `azure_ad_token_provider` is an exact name of a parameter used in `openai.AsyncAzureOpenAI` ([link](https://github.com/openai/openai-python/blob/798c6cb8032f4568d0b23f762d479d68be62ce3e/src/openai/lib/azure.py#L312)). Don't know how can I make the function `get_azure_ad_token_provider` less misleading, it basically gets the `azure_ad_token_provider` (the function that fetches the temporary token I mentioned above) required by `openai.AsyncAzureOpenAI`.

**Comment by kzych-inpost** [comment]
To add more context: advantage of this solution is that it passes the responsibility for Azure authentication to Azure SDK.

File: litellm/proxy/secret_managers/get_azure_ad_token_provider.py
Comment (ghost) [inline litellm/proxy/secret_managers/get_azure_ad_token_provider.py]: can we move the `azure.` imports to inside the function. This way it's only required, if this approach is used

File: litellm/router_utils/client_initalization_utils.py
Comment (ghost) [inline litellm/router_utils/client_initalization_utils.py]: can we type this correctly, so it's clear what the potential values could be here

File: litellm/router_utils/client_initalization_utils.py
Comment (ghost) [inline litellm/router_utils/client_initalization_utils.py:318]: can we add a basic unit test here - https://github.com/BerriAI/litellm/blob/04d69464e2c98f62569ef3a108bbc7aecc4ac8a1/litellm/tests/test_router_client_init.py#L17

with a mock test to make sure there's no future regressions

**Comment by ghost** [comment]
Hi @kzych-inpost looks good - left some comments re: refactoring code + Adding unit tests, before we can merge. thank you for this PR!

File: litellm/router_utils/client_initalization_utils.py
Comment (kzych-inpost) [reply litellm/router_utils/client_initalization_utils.py]: ✅

File: litellm/proxy/secret_managers/get_azure_ad_token_provider.py
Comment (kzych-inpost) [reply litellm/proxy/secret_managers/get_azure_ad_token_provider.py]: ✅

File: litellm/router_utils/client_initalization_utils.py
Comment (kzych-inpost) [reply litellm/router_utils/client_initalization_utils.py:318]: 1. OK, should I make another test-function next to this one with mocking environment variables (`AZURE_CLIENT_ID`, `AZURE_CLIENT_SECRET` and `AZURE_TENANT_ID`)?
2. The test you mentioned is skipped on my machine (problably due to the missing `pytest-asyncio` plugin and `pytest.mark.asyncio` decorator). Is this expected behaviour and this would be executed on CI/CD?
3. Can't test it locally anyway because I have no `redis` installed.

Comment (kzych-inpost) [reply litellm/router_utils/client_initalization_utils.py:318]: Prepared the commit [85a8b9e](https://github.com/BerriAI/litellm/pull/5131/commits/85a8b9ee3db14ed7537d7ff75807867e151f01dc) with environment variables and other prerequisites mocked.

**Comment by kzych-inpost** [comment]
How about the change in the documentation indicatiing that this feature is possible?

**Comment by kzych-inpost** [comment]
@krrishdholakia all requested changes are already commited :)
Are we ready to merge then?

**Pull request merged**
